res/res_pjsip.c

```diff
@@ -2727,6 +2864,603 @@ void ast_sip_modify_id_header(pj_pool_t *pool, pjsip_fromto_hdr *id_hdr, const s
+ *
+ * \note If the contact URI found for the endpoint already has a user in
+ * its URI, it will be replaced by the user passed as an argument to this function.
+ */
+static int insert_user_in_contact_uri(const char *to, const char *endpoint_name, const char *aors,
+	const char *user, char **uri)
+{
+	char *scheme = NULL;
+	char *contact_uri = NULL;
+	char *after_scheme = NULL;
+	char *host;
+	struct ast_sip_contact *contact = NULL;
+
+	contact = ast_sip_location_retrieve_contact_from_aor_list(aors);
+	if (!contact) {
+		ast_log(LOG_WARNING, "Dest: '%s'. Couldn't find contact for endpoint '%s'\n",
+			to, endpoint_name);
+		return -1;
+	}
+
+	contact_uri = ast_strdupa(contact->uri);
+	ao2_cleanup(contact);
+
+	ast_debug(3, "Dest: '%s' User: '%s'  Endpoint: '%s'  ContactURI: '%s'\n", to, user, endpoint_name, contact_uri);
+
+	/*
+	 * Contact URIs must have a scheme so we must insert the user between it and the host.
+	 */
+	scheme = contact_uri;
+	after_scheme = strchr(contact_uri, ':');
+	if (!after_scheme) {
+		/* A contact URI without a scheme?  Something's wrong.  Bail */
+		ast_log(LOG_WARNING, "Dest: '%s'. There was no scheme in the contact URI '%s'\n",
+			to, contact_uri);
+		return -1;
+	}
+	/*
+	 * Terminate the scheme.
+	 */
+	*after_scheme = '\0';
+	after_scheme++;
+
+	/*
+	 * If the contact_uri already has a user, the host starts after the '@', otherwise
+	 * the host is at after_scheme.
+	 *
+	 * We're going to ignore the existing user.
+	 */
+	host = strchr(after_scheme, '@');
+	if (host) {
+		host++;
+	} else {
+		host = after_scheme;
+	}
+
+	*uri = ast_malloc(strlen(scheme) + strlen(user) + strlen(host) + 3 /* One for the ':', '@' and terminating NULL */);
+	if (!(*uri)) {
+		return -1;
+	}
+
+	sprintf(*uri, "%s:%s@%s", scheme, user, host); /* Safe */
+
+	return 0;
+}
+
+/*!
+ * \internal
+ * \brief Get endpoint and URI when the destination is only a single token
+ *
+ * "destination" could be one of the following:
+ * \verbatim
+		endpoint_name
+		hostname
+ * \endverbatim
+ *
+ * \param to
+ * \param destination
+ * \param get_default_outbound If nonzero, try to retrieve the default
+ * 			       outbound endpoint if no endpoint was found.
+ * 			       Otherwise, return NULL if no endpoint was found.
+ * \param uri Pointer to URI variable.  Must be freed by caller - even if the return value is NULL!
+ * \return endpoint
+ */
+static struct ast_sip_endpoint *handle_single_token(const char *to, char *destination, int get_default_outbound, char **uri) {
+	char *endpoint_name = NULL;
+	struct ast_sip_endpoint *endpoint = NULL;
+	struct ast_sip_contact *contact = NULL;
+
+	/*
+	 * If "destination" is just one token, it could be an endpoint name
+	 * or a hostname without a scheme.
+	 */
+
+	endpoint = ast_sorcery_retrieve_by_id(ast_sip_get_sorcery(), "endpoint", destination);
+	if (!endpoint) {
+		/*
+		 * We can only assume it's a hostname.
+		 */
+		char *temp_uri = ast_malloc(strlen(destination) + strlen("sip:") + 1);
```

Comment:
Same applies here for uri.h

Everywhere that URI stuff is in this change should try to use the existing API instead.